src/rust/engine/fs/store/src/local.rs

```diff
@@ -144,12 +144,16 @@ impl UnderlyingByteStore for ShardedLmdb {
 }
 
 // We shard so there isn't a plethora of entries in one single dir.
+//
+// TODO: Add Arc'd inner struct to reduce clone costs.
 #[derive(Debug, Clone)]
 pub(crate) struct ShardedFSDB {
   root: PathBuf,
   executor: Executor,
   lease_time: Duration,
   dest_initializer: Arc<Mutex<HashMap<Fingerprint, Arc<OnceCell<()>>>>>,
+  // A cache of whether destination root directories are hardlinkable from the fsdb.
+  hardlinkable_destinations: Arc<Mutex<HashMap<PathBuf, Arc<OnceCell<bool>>>>>,
```

Comment:
What other root directories are there other than the sandbox root? I'm confused...

Reply:
There are at least:
1. the execution root - used for local process execution
2. the build root - used for `write_digest`

...but more generally, I don't want to bake in an assumption here that we are always materializing to a particular place.

Reply:
> the build root - used for write_digest

Does that mean we might hardlink into `dist`? If so I think that's a mistake, since I don't think we can guarantee people aren't gonna muck up `dist` :thinking:

Reply:
There is a separate check for that, iirc... the `force_mutable` flag is supposed to be used in that case.